### PR TITLE
Chinese Remainder Theorem | Diophantine Equation | Modular Division

### PR DESCRIPTION
**Pull request by OddExtension5**

### CODE REVIEW

**Comment by OddExtension5** [comment]
@cclauss Your review, please.

File: data_structures/hashing/number_theory/chinese_remainder_theorem.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+# If GCD(a,b) = 1, then for any remainder ra modulo a and any remainder rb modulo b there exists integer n,
+# such that n = ra (mod a) and n = ra(mod b).  If n1 and n2 are two such integers, then n1=n2(mod ab)
+
+# Algorithm :
+
+# 1. Use extended euclid algorithm to find x,y such that a*x + b*y = 1
+# 2. Take n = ra*by + rb*ax
+
+# import testmod for testing our function
+from doctest import testmod
```
Comment (cclauss) [inline data_structures/hashing/number_theory/chinese_remainder_theorem.py]: Let's move this line down just below line 68...
We do not need this import when we run your functions, only when we test them.

File: data_structures/hashing/number_theory/chinese_remainder_theorem.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    >>> chinese_remainder_theorem2(6,1,4,3)
+    14
+    """
+    x, y = invert_modulo(n1, n2), invert_modulo(n2, n1)
+    m = n1 * n2
+    n = r2 * x * n1 + r1 * y * n2
+    return (n % m + m) % m
+
+
+if __name__ == '__main__':
```
Comment (cclauss) [inline data_structures/hashing/number_theory/chinese_remainder_theorem.py]: from doctest import testmod

File: data_structures/hashing/number_theory/chinese_remainder_theorem.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+# Chinese Remainder Theorem:
+
+# If GCD(a,b) = 1, then for any remainder ra modulo a and any remainder rb modulo b there exists integer n,
```
Comment (cclauss) [inline data_structures/hashing/number_theory/chinese_remainder_theorem.py]: Please expand the acronym GCD and try to keep the lines to 88 characters or less.

File: data_structures/hashing/number_theory/chinese_remainder_theorem.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    m = n1 * n2
+    n = r2 * x * n1 + r1 * y * n2
+    return ((n % m + m) % m)
+
+
+# ----------SAME SOLUTION USING InvertModulo instead ExtendedEuclid----------------
+
+# This function find the inverses of a i.e., a^(-1)
+def invert_modulo(a, n):
+    (b, x) = extended_euclid(a, n)
```
Comment (cclauss) [inline data_structures/hashing/number_theory/chinese_remainder_theorem.py]: No doctest?

File: data_structures/hashing/number_theory/diophantine_equation.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        x = x0 + i * q
+        y = y0 - i * p
+        print(x, y)
+
+
+# Euclid's Lemma :  d divides a and b, if and only if d divides a-b and b
+
+# Euclid's Algorithm
+
+def gcd(a, b):
```
Comment (cclauss) [inline data_structures/hashing/number_theory/diophantine_equation.py]: Please expand the acronym gcd

File: data_structures/hashing/number_theory/modular_division.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+# Theorem:
+# a has a multiplicative inverse modulo n iff gcd(a,n) = 1
+
+
+# This find x = b*a^(-1) mod n
+# Uses ExtendedEuclid to find the inverse of a
+
+# Import testmod for testing our function
+from doctest import testmod
```
Comment (cclauss) [inline data_structures/hashing/number_theory/modular_division.py]: Move down to main()

**Comment by OddExtension5** [comment]
All Done!  @cclauss @cozek

**Comment by cclauss** [comment]
OK...
1. Please close  #1236 and #1232
2. Please change the path for these files from __data_structures/hashing/number_theory__ to __blockchain__.

**Comment by OddExtension5** [comment]
Done..

**Pull request merged**